src/cluster.c

```diff
@@ -1571,14 +1571,76 @@ void readonlyCommand(client *c) {
     addReply(c,shared.ok);
 }
 
-void replySlotsFlushAndFree(client *c, SlotsFlush *sflush) {
-    addReplyArrayLen(c, sflush->numRanges);
-    for (int i = 0 ; i < sflush->numRanges ; i++) {
+void replySlotsFlushAndFree(client *c, slotRangeArray *sra) {
+    addReplyArrayLen(c, sra->num_ranges);
+    for (int i = 0 ; i < sra->num_ranges ; i++) {
         addReplyArrayLen(c, 2);
-        addReplyLongLong(c, sflush->ranges[i].first);
-        addReplyLongLong(c, sflush->ranges[i].last);
+        addReplyLongLong(c, sra->ranges[i].start_slot);
+        addReplyLongLong(c, sra->ranges[i].end_slot);
     }
-    zfree(sflush);
+    zfree(sra);
+}
+
+/* Checks that slot ranges are well-formed and non-overlapping. */
+int validateSlotRanges(slotRangeArray *sra, sds *err) {
+    unsigned char *slots[CLUSTER_SLOTS] = {0};
+
+    for (int i = 0; i < sra->num_ranges; i++) {
+        if (sra->ranges[i].start_slot >= CLUSTER_SLOTS ||
+            sra->ranges[i].end_slot >= CLUSTER_SLOTS)
+        {
+            *err = sdscatprintf(sdsempty(), "slot range is out of range: %d-%d",
+                                sra->ranges[i].start_slot, sra->ranges[i].end_slot);
+            return C_ERR;
+        }
+
+        if (sra->ranges[i].start_slot > sra->ranges[i].end_slot) {
+            *err = sdscatprintf(sdsempty(), "start slot number %d is greater than end slot number %d",
+                                sra->ranges[i].start_slot, sra->ranges[i].end_slot);
+            return C_ERR;
+        }
+
+        for (int j = sra->ranges[i].start_slot; j <= sra->ranges[i].end_slot; j++) {
+            if (slots[j]) {
+                *err = sdscatprintf(sdsempty(), "Slot %d specified multiple times", j);
+                return C_ERR;
+            }
+            slots[j]++;
+        }
+    }
+    return C_OK;
+}
+
+/* Parse slot ranges from the command arguments. Returns NULL on error. */
+slotRangeArray *parseSlotRangesOrReply(client *c, int argc, int pos) {
+    int start_slot, end_slot, count;
+    slotRangeArray *sra;
+
+    serverAssert(pos <= argc);
+    serverAssert((argc - pos) % 2 == 0);
+
+    count = (argc - pos) / 2;
+    sra = zcalloc(sizeof(*sra) + count * sizeof(slotRange));
```

Comment:
maybe we can init `num_ranges` explicitly

Reply:
you mean something like `sra->num_ranges = count` ? but later we use `sra->num_ranges` as index in the loop.